node/libs/roles/src/attester/tests.rs

```diff
@@ -0,0 +1,234 @@
+    let t = TextFmt::encode(&sk);
+    assert_eq!(
+        sk.public(),
+        Text::new(&t).decode::<SecretKey>().unwrap().public()
+    );
+
+    let pk: PublicKey = rng.gen();
+    let t = TextFmt::encode(&pk);
+    assert_eq!(pk, Text::new(&t).decode::<PublicKey>().unwrap());
+
+    let sig: Signature = rng.gen();
+    let t = TextFmt::encode(&sig);
+    assert_eq!(sig, Text::new(&t).decode::<Signature>().unwrap());
+
+    let agg_sig: AggregateSignature = rng.gen();
+    let t = TextFmt::encode(&agg_sig);
+    assert_eq!(
+        agg_sig,
+        Text::new(&t).decode::<AggregateSignature>().unwrap()
+    );
+
+    let msg_hash: MsgHash = rng.gen();
+    let t = TextFmt::encode(&msg_hash);
+    assert_eq!(msg_hash, Text::new(&t).decode::<MsgHash>().unwrap());
+}
+
+#[test]
+fn test_schema_encoding() {
+    let ctx = ctx::test_root(&ctx::RealClock);
+    let rng = &mut ctx.rng();
+    test_encode_random::<SignedBatchMsg<L1Batch>>(rng);
+    test_encode_random::<L1BatchQC>(rng);
+    test_encode_random::<Msg>(rng);
+    test_encode_random::<MsgHash>(rng);
+    test_encode_random::<Signers>(rng);
+    test_encode_random::<PublicKey>(rng);
+    test_encode_random::<Signature>(rng);
+    test_encode_random::<AggregateSignature>(rng);
+}
+
+#[test]
+fn test_signature_verify() {
+    let ctx = ctx::test_root(&ctx::RealClock);
+    let rng = &mut ctx.rng();
+
+    let msg1: MsgHash = rng.gen();
+    let msg2: MsgHash = rng.gen();
+
+    let key1: SecretKey = rng.gen();
+    let key2: SecretKey = rng.gen();
+
+    let sig1 = key1.sign_hash(&msg1);
+
+    // Matching key and message.
+    sig1.verify_hash(&msg1, &key1.public()).unwrap();
+
+    // Mismatching message.
+    assert!(sig1.verify_hash(&msg2, &key1.public()).is_err());
+
+    // Mismatching key.
+    assert!(sig1.verify_hash(&msg1, &key2.public()).is_err());
+}
+
+#[test]
+fn test_agg_signature_verify() {
+    let ctx = ctx::test_root(&ctx::RealClock);
+    let rng = &mut ctx.rng();
+
+    let msg1: MsgHash = rng.gen();
+    let msg2: MsgHash = rng.gen();
+
+    let key1: SecretKey = rng.gen();
+    let key2: SecretKey = rng.gen();
+
+    let sig1 = key1.sign_hash(&msg1);
+    let sig2 = key2.sign_hash(&msg2);
+
+    let agg_sig = AggregateSignature::aggregate(vec![&sig1, &sig2]);
+
+    // Matching key and message.
+    agg_sig
+        .verify_hash([(msg1, &key1.public()), (msg2, &key2.public())].into_iter())
+        .unwrap();
+
+    // Mismatching message.
+    assert!(agg_sig
+        .verify_hash([(msg2, &key1.public()), (msg1, &key2.public())].into_iter())
+        .is_err());
+
+    // Mismatching key.
+    assert!(agg_sig
+        .verify_hash([(msg1, &key2.public()), (msg2, &key1.public())].into_iter())
+        .is_err());
+}
+
+fn make_l1_batch_msg(rng: &mut impl Rng) -> L1Batch {
+    L1Batch {
+        number: BatchNumber(rng.gen()),
+        timestamp: time::UNIX_EPOCH + time::Duration::seconds(rng.gen_range(0..1000000000)),
+    }
+}
```

Comment:
isn't it the same as rng.gen::<L1Batch>() ?

Reply:
I believe so, the timestamp will no longer be used anyway.